Number_types/include/CGAL/boost_mp.h

```diff
@@ -139,6 +141,259 @@ struct Algebraic_structure_traits<boost::multiprecision::detail::expression<T1,T
 
 // Real_embeddable_traits
 
+namespace Boost_MP_internal {
+
+  // here we know that `intv` contains int64 numbers such that their msb is std::numeric_limits<double>::digits-1
+  inline
+  Interval_nt<false> shift_positive_interval( const Interval_nt<false>& intv, const int e ) {
+    CGAL_assertion(intv.inf() > 0.0);
+    CGAL_assertion(intv.sup() > 0.0);
+    typedef std::numeric_limits<double> limits;
+
+    if (e < std::numeric_limits<double>::min_exponent)
+    {
+      if ( e+std::numeric_limits<double>::digits < std::numeric_limits<double>::min_exponent)
+        return CGAL::Interval_nt<false>(0, (limits::min)());
+      // following calls to ldexp call are exact (e+digits is larger or equal to min_exponent)
+      return CGAL::Interval_nt<false>(std::ldexp(intv.inf(), e), std::ldexp(intv.sup(), e));
```

Comment:
If those are exact, then so is `ldexp(1,e)`, why do you need this special case?

Reply:
not if `e < std::numeric_limits<double>::min_exponent`

Reply:
Er, there are denormal values below that, we can use them...

Reply:
But mostly, the comment says that `ldexp(intv.inf(),e)` is exact without having looked at `intv`, so the comment really seems to imply that `ldexp(1,e)` is exact, since 1 is a possible value for `intv.inf()`. No?

Reply:
The thing is I was suspecting that `ldexp` is not exact as soon as you switch to denormals, especially with MSVC. 
About `intv.inf()`, it cannot be 1 as `msb()` is 52. However, it might be that `msb()` of `intv.sup()` is 53 since it is `inf()+1` in case the division is not exact.

Reply:
I thought the problem with MSVC was the rounding mode, I would expect it to handle denormals correctly in the exact case, but I didn't try.